README.md
# Baryonyx-0.2.1

[![Build Status](https://travis-ci.org/quesnel/baryonyx.png?branch=master)](https://travis-ci.org/quesnel/baryonyx)

[Baryonyx](https://en.wikipedia.org/wiki/Baryonyx) is an integer and
binary linear programming solver based on
the [Dag Wedelin](http://www.cse.chalmers.se/~dag/) heuristic.

Copyright © 2017 [INRA](http://www.inra.fr/en)

Author [Gauthier Quesnel](https://mia.toulouse.inra.fr/Gauthier_QUESNEL) <gauthier.quesnel@inra.fr>

The software is released under the MIT license. See the LICENSE file.

## Requirements

* `cmake` (≥ 3)
* c++ compiler with c++14 support (`gcc` ≥ 5, `clang` ≥ 3.5)

For recent Debian and Ubuntu derivatives (remove clang to only use
gcc):

    apt-get install build-essential cmake clang

## Installation

Default, baryonyx provides a shared library `libbaryonyx-0.2.so`, a static
library `libbaryonyx-0.2.a` and an executable `baryonyx-0.2`. To compile and
install in default cmake's install directory:

    cd baryonyx
    mkdir build
    cd build
    cmake -DCMAKE_BUILD_TYPE=Release ..
    make install

Previous command line install baryonyx program and library into the
`/usr/local` prefix. If you install into another directory, you need to define
three environment variables (into your `.bashrc` or equivalent). If you install
into `$HOME/usr` for example, you need to define in your `.bashrc`:

    export PATH=$PATH:$HOME/usr
    export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$HOME/usr/lib
    export PKG_CONFIG_PATH=$PKG_CONFIX_PATH:$HOME/usr/lib/pkgconfig

Then run the following commands:

    cd baryonyx
    mkdir build
    cd build
    cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=$HOME/usr ..
    make install

To override the default build flags, remove the `CMAKE_BUILD_TYPE`
parameter and override the `CXXFLAGS` as follow:

    export CC=gcc-7
    export CXX=g++-7
    export CXXFLAGS='-Wall -Wextra -Werror -ftree-vectorize -mmmx -msse -msse2 -msse3 -O3'
    cmake -DCMAKE_INSTALL_PREFIX=$HOME/usr ..

## Usage

To run baryonyx in solver mode (ie. trying to valid all constraints):

    baryonyx-0.2 file.lp

To run baryonyx into the heuristic model (ie. trying to valid all constraints
and optimize the solution), add a `-o` or `--optimize` option to the command
line:

    baryonyx-0.2 file.lp

The baryonyx solver have many parameters. To assign it, use the `-p
[name]:value` syntax in the command line:

* `real` **time_limit**: time in second to stop the solver or the heuristic
* `real` **theta**:
* `real` **delta**:
* `real` **kappa_min**:
* `real` **kappa_step**:
* `real` **kappa_max**:
* `real` **alpha**:
* `real` **pushing_k_factor**:
* `integer` **pushes_limit**:
* `integer` **pushing_objective_amplifier**:
* `integer` **pushing_iteration_limit**:
* `integer` **limit**: number of loop to stop the solver or the heuristic
* `integer` **w**: warmup-iterator (number of loop without updating kappa)
* `constraint_order` **order**: `none`, `reversing`, `random-sorting`, `infeasibility-decr`, `infeasibility-incr`.
* `string` **preprocessing**: For constraints matrix A, add new static order
  for constraints (`-p preprocessing=option`) where option is defined as:
** `none` (default): use the lp format constraint order.
** `less,greater,equal` or `greater,less,equal` or any other form: use the type
   of operator
** `variables-number` (experimental): use number variable links.
** `variables-weight` (experimental): sort variables weights (sum variable
   appearance in all constraints).
** `constraints-weight` (experimental): use number constraint links.
* `string` **norm**: `none`, `l1`, `l2`, `rng`, `infinity`
* `integer` **print-level**: show debug information if greater than 0.

For example:

    baryonyx -p limit:1000000 lib/test/prevl1.lp
    baryonyx -p limit: -p kappa-min:0.2 lib/test/prevl1.lp

## Upgrade

To upgrade to the latest version of baryonyx, under bash (or equivalent):

    cd baryonyx
    git pull -r
    cd build
    make install

# R

To use rbaryonyx, you must compile and install the baryonyx library. Follow
the previous section and install R.

## Installation

To install rbaryonyx, first, remove old installation under bash (or
equivalent):

    cd baryonyx/rbaryonyx
    # Removes previous installed version of rbaryonyx
    R CMD REMOVE rbaryonyx

The R rbaryonyx package requires several packages. Then, under a R terminal:

    # Install the dependencies of rbaryonyx
    install.packages("roxygen2")
    install.packages("Rcpp")
    install.packages("devtools")

    library(Rcpp)
    compileAttributes(".")
    library(devtools)
    devtools::document()
    devtools::build()
    devtools::install()

    library(rbaryonyx)
    ?rbaryonyx

## API

Two functions are provided to solve or optimize 01 linear programming problem. Parameters are the same as `C++ API`. These function returns a scalar:

- If a solution is found:
  - if the problem is a minimization: the value of the solution found.
  - if the problem is a maximization: the inverse of the solution found.
- If no solution is found, we use the limits of the objective
  function (minimal and maximal value possible.
  - if the problem is a minimization: the maximal value possible + the remaining constraints.
  - if the problem is a maximization: the inverse of the minimal value possible + the remaining constraints.
- If a error occurred (not enough memory, problem error etc.):
  - if the problem is a minimization: the maximal value possible + the number of constraints .
  - if the problem is a maximization: the inverse of the minimal value possible + the number of constraints.

```R
solve_01lp_problem <- function(file_path, limit = 1000L, theta = 0.5,
  delta = 1e-4, constraint_order = 0L, kappa_min = 0.1, kappa_step = 1e-4,
  kappa_max = 1.0, alpha = 1.0, w = 500L, time_limit = 10.0, seed = -1L,
  thread = 1L, norm = 4L, pushing_k_factor = 0.9,
  pushing_objective_amplifier = 5.0, pushes_limit = 10L,
  pushing_iteration_limit = 20L, float_type = 1L, verbose = TRUE)

optimize_01lp_problem <- function(file_path, limit = 1000L, theta = 0.5,
  delta = 1e-4, constraint_order = 0L, kappa_min = 0.1, kappa_step = 1e-4,
  kappa_max = 1.0, alpha = 1.0, w = 500L, time_limit = 10.0, seed = -1L,
  thread = 1L, norm = 4L, pushing_k_factor = 0.9,
  pushing_objective_amplifier = 5.0, pushes_limit = 10L,
  pushing_iteration_limit = 20L, float_type = 1L, verbose = TRUE)

```

## Usage

Apply morris method to found useful parameters:

```R
library(rbaryonyx)
library(sensitivity)

factors = c("theta", "delta", "constraint_order", "kappa_min", "kappa_step",
  "kappa_max", "alpha", "w", "norm", "pushing_k_factor",
  "pushing_objective_amplifier", "pushes_limit", "pushing_iteration_limit",
  "float_type")

bounds = data.frame(
  min=c(
    0,     # theta
    0,     # delta
    0,     # constraint_order
    0,     # kappa_min
    1e-16, # kappa_step
    1.0,   # kappa_max
    0.0,   # alpha
    50,    # w
    0,     # norm
    0.1,   # pushing_k_factor
    1.0,   # pushing_objective_amplifier
    10,    # pushes_limit
    20,    # pushing_iteration_limit
    0
    ),    # float_type
max=c(
    1,     # theta
    0,     # delta
    4,     # constraint_order
    0.1,   # kappa_min
    1e-1,  # kappa_step
    1.0,   # kappa_max
    2.0,   # alpha
    500,   # w
    4,     # norm
    1,     # pushing_k_factor
    10.0,  # pushing_objective_amplifier
    100,   # pushes_limit
    200,   # pushing_iteration_limit
    2))    # float_type

rownames(bounds) <- factors

morrisDesign <- morris(model = NULL,
                factors = factors,
                r = 10,
                design=list(type="oat", levels=10, grid.jump=5),
                binf = bounds$min,
                bsup = bounds$max,
                scale=TRUE)

solve_lp <- function(x, file_path, limit=10000, time_limit=10, seed=123456789, thread=1) {
  r <- rbaryonyx::solve_01lp_problem(file_path = file_path,
                   limit = limit,
                   theta = x["theta"],
                   delta = x["delta"],
                   constraint_order = x["constraint_order"],
                   kappa_min = x["kappa_min"],
                   kappa_step = x["kappa_step"],
                   kappa_max = x["kappa_max"],
                   alpha = x["alpha"],
                   w = x["w"],
                   time_limit = time_limit,
                   seed = seed,
                   thread = thread,
                   norm = x["norm"],
                   pushing_k_factor = x["pushing_k_factor"],
                   pushing_objective_amplifier = x["pushing_objective_amplifier,"],
                   pushes_limit = x["pushes_limit"],
                   pushing_iteration_limit = x["pushing_iteration_limit"],
                   float_type = x["float_type"])

  return(r)
}

r = apply(morrisDesign$X, 1, solve_lp, file_path="verger_5_5.lp", thread=1, limit=10000, time_limit=10, seed=123456789)

morrisDesign$Y <- r
mu <- apply(morrisDesign$X,2,mean)
mu.star <- apply(morrisDesign$X, 2, function(x) mean(abs(x)))
sigma <- apply(morrisDesign$ee, 2, sd)

apply(morrisDesign$X, 2, function(v) plot(factor(v), r))
```

# Upgrade

To upgrade to the latest version of rbaryonyx, under bash (or equivalent):

```bash
cd baryonyx
git pull -r
cd build
make install
R CMD REMOVE rbaryonyx
cd rbaryonyx
Rscript -e 'library(Rcpp); compileAttributes(".")'
Rscript -e 'library(devtools); devtools::document()'
cd ..
R CMD build rbaryonyx
R CMD INSTALL rbaryonyx_1.0.tar.gz
```


NEWS
## baryonyx v0.5.0 - 2020-02-04

### Major changes:

* project:
  * add [codecov.io](https://codecove.io) to produce coverage reports in travis
  * use Visual Studio 2019 image in appveyor
  * [boost/ut](https://github.com/boost-experimental/ut): use boost.UT v1.1.6
  * [fmtlib/fmt](https://fmt.dev/latest/index.html): bump to version v6.1.1

* optimizer:
  * use a evolutionary algorithm to init/reinit solver. In optimization mode, Baryonyx stores all solutions found during search. Two algorithms are used to initialize the X vector:
    * A crossover/mutation using two solutions is used to re-initialize the solver.
    * An improve mode using a dynamical kappa.
  * Quadratic objective function: Experimental support.
  * Z-coefficient: Enable the Z-coefficient solver.
  * random-solver: This solver is enabled with the `--random` command line option
  and in the `solver_parameters::solver_type`. It provides a pure random solver without any preference and Lagrangian relaxation. It is provided only for benchmark.
  * add Lagrangian multipliers sign change detection.

### Small changes:

* c++:
  * solvers: replace `fill` and `copy` with `fill_n` and `copy_n` equivalent.
  * solvers: replace `std::vector<T>` with a new `bit_array` to improve cache  efficiency
  * main: fix signed/unsigned warning

* solver
  * negative coefficient solvers: fix R access
  * add pi signed change detection
  * improve efficiency of random-solver
  * add reset function to reinit solver
  * solver: move normalized cost in thread part

* optimizer:
  * replace time limit detection algorithm
  * add pi signed change detection to build a new constraint remaining list

* benchmark:
  * fix sort functor behaviour

* main:
  * fix time-limit parameter assignation
  * missing or malformed parameters equal error
  * fix infinity norm string
  * add a `--debug` mode.

* CMake:
  * reduce warning level for msvc
  * use the BUILD_SHARED_LIBS CMake option to build share library. Default, a
     static library is build to enable all internal test that use private API.

## baryonyx v0.4.0 - 2019-02-19

Bump C++ dependency to C++17 standard.

main::

* benchmark: add a second output `csv` file with the rank for each solvers and
  models to show best solver.

solver/optimizer::

* Optimize all solver algorithms to improve performance: rewrite and
  generalize.

* Improve the `best-cycle` algorithm: is no best solution was found for three
  complete optimization, solver restart to `bastert-cycle`.

* Fix the initialization cycle behaviour.

* Fix the random normalization behaviour.

* Remove useless header (sparse-vector etc), dead code.

== baryonyx v0.3.4 - 2019-01-17

main::

* benchmark: add a benchmark mode to permit to run benchmark on a set of
  problems described in a `csv` files and make easy comparison.

* colours: improve colour rendering in log system. Replace specific Linux/Unix
  source code about colour in (pseudo)terminal with the `fmt::text_style` and
  `fmt::terminal_color` enumeration.

* add warning message for unknown parameters

solver/optimizer::

* core: add init policy cycle:
  ** `bastert` (default): use cost values.
  ** `random`: use random value.
  ** `best`: use best solution found if available (`bastert` otherwise).
  ** `bastert-cycle`: start with bastert policy then start a cycle between
       random, best and bastert.
  ** `random-cycle`: start with random policy then start a cycle between best,
       bastert and random.
  ** `best-cycle`: start with best policy then start a cycle between bastert,
       random and best.

* core: add lagrangian runtime constraint order.
  ** `lagrangian-decr`: sort violated constraints according to the Lagrangian
     multiplier values in decremental order.
  ** `lagrangian-incr`: sort violated constraints according to the Lagrangian
     multiplier values in incremental order

## baryonyx v0.3.3 - 2018-10-04

main::

* output: better render output float number

solver/optimizer::

* Fix the preference matrix assignment for both equalities and inequalities
  with -1 and 1 coefficients, we replace the preference matrix clean up
  algorithm (invert the variable values for negative coefficients) with a
  version that revert only the added or subtracted value during the preferences
  update process.
* Fix core dump if no solution are found

preprocessor::

* update the preprocessing solver option: Replace the old preprocessing option
  with new one but without description (p1, p2, p3 and p4). Options are `none`
  to use the raw_problem (or lp file) order for constraints and variables.
  `memory`, the default, to use the raw_problem (or lp file) order for
  constraints but sort the variables to improve the memory cache efficiency.

## baryonyx v0.3.2 - 2018-09-24

solver/optimizer::

* Be sure to always return the correct solution value (if any), the
  status (success, time-limit, kappa-limit, loop-limit) and the
  remaining constraints.
* Fix optimizer algorithm to store best constraints remaining
  solution.
* Fix push system in solver

general::

* add storage policy option to select the result solution vector
  policy (one best solution, best and worst solution and the five best
  solution).
* bump fmt library to v5.2.0

preprocessor::

* detect and affect variable affected in bound.

main::

* Add the `storage-type` option with a string parameter (one, bound
  and five).

## baryonyx v0.3.1 - 2018-09-18

main::

* fix typo in constraint-order parameter.
* fix exception output message.

preprocessor::

* replace the old preprocessor algorithm. The new one is better to remove
  constraints and variables and fix some issues.

solver::

* print solver parameters before solve/optimize

parser::

* fix the reading of the constraint label.
* add a default label to constraints.
* remove function elements where factors equals 0.
* merge function elements in constraints or in the objective function where
  variable index appears several times.

rilp/src/rilp.cpp

/* Copyright (C) 2017 INRA
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the
 * "Software"), to deal in he Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#define R_USE_C99_IN_CXX
#include <Rcpp.h>

#include <lpcore>

using namespace Rcpp;

class Rcontext : public lp::context::logger
{
public:
    Rcontext() = default;
    ~Rcontext() noexcept override = default;

    void write(int priority,
               const char* file,
               int line,
               const char* fn,
               const char* format,
               va_list args) noexcept override final
    {
        Rprintf("lp: %d at %d in function %s form file %s: ",
                priority,
                line,
                fn,
                file);

        Rvprintf(format, args);
    }

    void write(lp::context::message_type m,
               const char* format,
               va_list args) noexcept override final
    {
        switch (m) {
        case lp::context::message_type::emerg:
            Rprintf("lp- System is unusable: ");
            break;
        case lp::context::message_type::alert:
            Rprintf("lp- Action must be taken immediately: ");
            break;
        case lp::context::message_type::crit:
            Rprintf("lp- critical conditions: ");
            break;
        case lp::context::message_type::err:
            Rprintf("lp- error conditions: ");
            break;
        case lp::context::message_type::warning:
            Rprintf("lp- warning conditions: ");
            break;
        case lp::context::message_type::notice:
        case lp::context::message_type::info:
        case lp::context::message_type::debug:
            break;
        }

        Rvprintf(format, args);
    }
};

void
assign_parameters(std::shared_ptr<lp::context> ctx,
                  long int limit,
                  double theta,
                  double delta,
                  long int constraint_order,
                  double kappa_min,
                  double kappa_step,
                  double kappa_max,
                  double alpha,
                  long int w,
                  double time_limit,
                  long int seed,
                  long int thread)
{
    ctx->set_parameter("limit", limit);
    ctx->set_parameter("theta", theta);
    ctx->set_parameter("delta", delta);

    switch (constraint_order) {
    case 1:
        ctx->set_parameter("constraint-order", std::string("reversing"));
        break;
    case 2:
        ctx->set_parameter("constraint-order", std::string("random-sorting"));
        break;
    case 3:
        ctx->set_parameter("constraint-order",
                           std::string("infeasibility-decr"));
        break;
    case 4:
        ctx->set_parameter("constraint-order",
                           std::string("infeasibility-incr"));
        break;
    case 0:
    default:
        ctx->set_parameter("constraint-order", std::string("none"));
        break;
    }

    ctx->set_parameter("kappa-min", kappa_min);
    ctx->set_parameter("kappa-step", kappa_step);
    ctx->set_parameter("kappa-max", kappa_max);
    ctx->set_parameter("alpha", alpha);
    ctx->set_parameter("w", w);
    ctx->set_parameter("time-limit", time_limit);

    if (seed > 0)
        ctx->set_parameter("seed", seed);

    ctx->set_parameter("thread", thread);
}

//' Tries to solve the 01 linear programming problem.
//'
//' @param constraint_order: 0-none, 1-reversing, 2-random-sorting,
//' 3-infeasibility-decr, 4- infeasibility-incr
//'
//' @return A list of three values: The remaining constraint and the value
//' of the solution (Integer, Real, Real):
//'   - Integer: The number of remaining constraints. 0 means a solution
//'      found, greater than 0 means remaining constraints and value is
//'      0. -1 means an error occurred during solving. Use the @c `lp`
//'      program to check out the error.
//'   - Real: The value of the solution found (if the remaining
//'     constraint equal 0).
//'   - Real: The value [objective_function_min, objective_function_max +
//'     constraints number].
//'
//' @useDynLib rilp
//' @importFrom Rcpp sourceCpp
//'
//' @export
// [[Rcpp::export]]
List
solve_01lp_problem(std::string file_path,
                   long int limit = 1000,
                   double theta = 0.5,
                   double delta = 1e-4,
                   long int constraint_order = 0,
                   double kappa_min = 0.1,
                   double kappa_step = 1e-4,
                   double kappa_max = 1.0,
                   double alpha = 1.0,
                   long int w = 60,
                   double time_limit = 10.0,
                   long int seed = -1,
                   long int thread = 1,
                   bool verbose = true) noexcept
{
    try {
        auto ctx = std::make_shared<lp::context>();
        ctx->set_logger(std::make_unique<Rcontext>());
        ctx->set_log_priority(!verbose ? 6 : 3);

        auto pb = lp::make_problem(ctx, file_path);
        auto mm = lp::compute_min_max_objective_function(pb);

        if (pb.type == lp::objective_function_type::maximize) {
            std::swap(std::get<0>(mm), std::get<1>(mm));
            std::get<0>(mm) = -std::get<0>(mm);
            std::get<1>(mm) = -std::get<1>(mm);
        }

        assign_parameters(ctx,
                          limit,
                          theta,
                          delta,
                          constraint_order,
                          kappa_min,
                          kappa_step,
                          kappa_max,
                          alpha,
                          w,
                          time_limit,
                          seed,
                          thread);

        auto result = lp::solve(ctx, pb);

        if (result.status == lp::result_status::success) {
            if (pb.type == lp::objective_function_type::maximize)
                return List::create(
                  result.remaining_constraints, result.value, -result.value);
            else
                return List::create(
                  result.remaining_constraints, result.value, result.value);
        }

        return List::create(result.remaining_constraints,
                            result.value,
                            std::get<1>(mm) + result.remaining_constraints);
    } catch (const std::bad_alloc& e) {
        Rprintf("lp memory error: %s\n", e.what());
    } catch (const std::exception& e) {
        Rprintf("lp error: %s\n", e.what());
    } catch (...) {
        Rprintf("lp error: unknown error\n");
    }

    return List::create(
      NA_INTEGER, NA_REAL, std::numeric_limits<double>::infinity());
}

//' Tries to optimize the 01 linear programming problem.
//'
//' @param constraint_order: 0-none, 1-reversing, 2-random-sorting,
//' 3-infeasibility-decr, 4- infeasibility-incr
//'
//' @return an uniq scalar.
//'   - If a solution is found:
//'     - if the problem is a minimization: the value of the solution found.
//'     - if the problem is a maximization: the inverse of the solution
//'       found.
//'   - If no solution is found, we use the limits of the objective
//'     function (minimal and maximal value possible.
//'     - if the problem is a minimization: the maximal value possible +
//'       the remaining constraints.
//'     - if the problem is a maximization: the inverse of the minimal
//'       value possible + the remaining constraints.
//'   - If a error occurred (not enough memory, problem error etc.):
//'     - if the problem is a minimization: the maximal value possible +
//'       the number of constraints .
//'     - if the problem is a maximization: the inverse of the minimal
//'       value possible + the number of constraints.
//'
//' @useDynLib rilp
//' @importFrom Rcpp sourceCpp
//'
//' @export
// [[Rcpp::export]]
double
optimize_01lp_problem(std::string file_path,
                      long int limit = 1000,
                      double theta = 0.5,
                      double delta = 1e-4,
                      long int constraint_order = 0,
                      double kappa_min = 0.1,
                      double kappa_step = 1e-4,
                      double kappa_max = 1.0,
                      double alpha = 1.0,
                      long int w = 60,
                      double time_limit = 10.0,
                      double pushing_k_factor = 0.9,
                      long int pushes_limit = 50,
                      double pushing_objective_amplifier = 10,
                      long int pushing_iteration_limit = 10,
                      long int seed = -1,
                      long int thread = 1,
                      bool verbose = true) noexcept
{
    double ret{ 0 };

    try {
        auto ctx = std::make_shared<lp::context>();
        ctx->set_logger(std::make_unique<Rcontext>());
        ctx->set_log_priority(verbose ? 6 : 3);

        auto pb = lp::make_problem(ctx, file_path);
        auto mm = lp::compute_min_max_objective_function(pb);

        if (pb.type == lp::objective_function_type::maximize)
            ret = -std::get<0>(mm) + size(pb);
        else
            ret = std::get<1>(mm) + size(pb);

        if (pb.type == lp::objective_function_type::maximize) {
            std::swap(std::get<0>(mm), std::get<1>(mm));
            std::get<0>(mm) = -std::get<0>(mm);
            std::get<1>(mm) = -std::get<1>(mm);
        }

        assign_parameters(ctx,
                          limit,
                          theta,
                          delta,
                          constraint_order,
                          kappa_min,
                          kappa_step,
                          kappa_max,
                          alpha,
                          w,
                          time_limit,
                          seed,
                          thread);

        ctx->set_parameter("pushing-k-factor", pushing_k_factor);
        ctx->set_parameter("pushes-limit", pushes_limit);
        ctx->set_parameter("pushing-objective-amplifier",
                           pushing_objective_amplifier);
        ctx->set_parameter("pushing-iteration-limit", pushing_iteration_limit);

        auto result = lp::optimize(ctx, pb);

        if (result.status == lp::result_status::success) {
            if (pb.type == lp::objective_function_type::maximize)
                ret = -result.value;
            else
                ret = result.value;
        } else {
            if (pb.type == lp::objective_function_type::maximize)
                ret = -std::get<0>(mm) + result.remaining_constraints;
            else
                ret = std::get<1>(mm) + result.remaining_constraints;
        }
    } catch (const std::bad_alloc& e) {
        Rprintf("lp memory error: %s\n", e.what());
    } catch (const std::exception& e) {
        Rprintf("lp error: %s\n", e.what());
    } catch (...) {
        Rprintf("lp error: unknown error\n");
    }

    Rprintf("optimizer returns: %f (kappa: %f %f %f delta: %f theta: %f)\n",
            ret,
            kappa_min,
            kappa_step,
            kappa_max,
            delta,
            theta);

    return ret;
}
